### DISCUSSION TITLE
Is it acceptable to `extend` user supplied instances of logger?

### DISCUSSION THREAD

**Discussion by ioquatix** [sidekiq/sidekiq#5673] [General]
https://github.com/mperham/sidekiq/blob/c4dbc922305effd1db98527c293e0354456fc79a/lib/sidekiq/config.rb#L251

This line is causing a lot of issues when assigning instances which aren't instances of `::Logger` or that fake the `::Logger` interface.

Is there a better way to do this?

Simply doing `instance_variable_set(:@logger, fake_logger_interface)` works correctly.

Is per-job log level really that important?

**Comment by mperham** [comment]
I'm open to refactoring which improves those scenarios.

**Reply by ioquatix** [reply]
Okay, so:

1. Remove it.
2. Make an argument to control it.
3. Something else?

What do you prefer?

My understanding of this monkey patch is to make per-job log levels. I'm not sure why `Sidekiq.logger = logger` should alter logger at all, and from that point on makes log level per-fiber which seems like a bug TBH. Altering the user supplied object using `extend` seems highly invasive.

I'd personally vote for (1), because the default logger already includes the monkey patch in a compatible way, which makes sense to me. Alternatively you could append `if logger.is_a?(::Logger)` to only apply the monkey patch if it's guaranteed to work (which it isn't since you are assuming the internals of the logger gem which could in theory change).

**Reply by mperham** [reply]
I think it was pitched as useful when you want to debug a specific type of job in production you don’t need to turn on the firehose for all jobs but rather target DEBUG level to a single instance you create via console. I can definitely see myself using that feature but if you can think of a reasonable alternative I’m not tied to this at all.

**Reply by ioquatix** [reply]
Well, at least this is the first thing I'd do: https://github.com/mperham/sidekiq/pull/5674 but even that is unsafe since you are depending on the internal implementation of `::Logger`, but it's better than nothing.

Then, as I said above, I'd suggest you remove it. Your use case does not change since the default `Sidekiq.logger` is an instance of `Sidekiq::Logger` which supports the per-fiber log level.

This only impacts people assigning custom loggers, and you are assuming they will be compatible with your mixin (which is an incorrect assumption). And in that case, they probably aren't expecting that behaviour anyway.

https://github.com/mperham/sidekiq/blob/main/lib/sidekiq/config.rb#L234-L243

By the way, lazy initialisation of a logger using `||=` is likely to be thread unsafe without a mutex.

**Reply by mperham** [reply]
> By the way, lazy initialisation of a logger using ||= is likely to be thread unsafe without a mutex.

The logger is touched on the main thread, before any other threads are created.

**Reply by ioquatix** [reply]
So why bother creating it lazily if that's the case?

**Reply by mperham** [reply]
So I can log anywhere anytime during Sidekiq bootup without having to worry "Have I initialized the logger yet?"

**Reply by ioquatix** [reply]
I don't really understand. Logger is an instance variable of `class Config`. Surely adding `@logger = logger || Logger.new...` to `Config#initialize` would be before any usage, since before that, the instance of config wouldn't even exist.

**Comment by ioquatix** [comment]
It was fixed, details in <https://github.com/mperham/sidekiq/pull/5674>.
